File: cp932_to_utf16.java
import java.lang.*;
import java.nio.*;
import java.nio.charset.*;

class cp932_to_utf16 {
	static int decode(int cp932) {
		//if (0xf040 <= cp932 && cp932 <= 0xf9fc) { // cp932 private use area
		//	return -1;
		//}

		ByteBuffer inbuf;
		if (cp932 < 0x100) {
			inbuf = ByteBuffer.allocate(1);
		} else {
			inbuf = ByteBuffer.allocate(2);
			inbuf.put((byte)((cp932 >> 8) & 0xff));
		}
		inbuf.put((byte)(cp932 & 0xff));
		inbuf.rewind();

		CharBuffer outbuf = CharBuffer.allocate(100);
		CharsetDecoder decoder = Charset.forName("MS932").newDecoder();
		decoder.reset();
		decoder.onMalformedInput(CodingErrorAction.REPORT);
		decoder.onUnmappableCharacter(CodingErrorAction.REPORT);
		CoderResult cr = decoder.decode(inbuf, outbuf, true);
		if (cr.isError() || cr.isOverflow()) {
			return -1;
		}
		cr = decoder.flush(outbuf);
		if (cr.isError() || cr.isOverflow() || outbuf.position() != 1) {
			return -1;
		}
		return outbuf.get(0);
	}
	
	public static void main(String[] args) {
		for (int cp932 = 0; cp932 <= 0xffff; ++ cp932) {
			int unicode = decode(cp932);
			if (0 <= unicode) {
				if (cp932 < 0x100) {
					System.out.printf("0x%02X\t", cp932);
				} else {
					System.out.printf("0x%04X\t", cp932);
				}
				System.out.printf("0x%04X\n", unicode);
			}
		}
	}
}

File: cp932_to_utf16.php
#!/usr/bin/php
<?php

function decode($cp932) {
	#if (0xf040 <= $cp932 && $cp932 <= 0xf9fc) { # $cp932 private use area
	#	return -1;
	#}
  
	if ($cp932 < 0x100) {
		$inbuf = sprintf("%c", $cp932 & 0xff);
	} else {
		$leadbyte = ($cp932 >> 8) & 0xff;
		if (!((0x81 <= $leadbyte && $leadbyte <= 0x9f) ||
			  (0xE0 <= $leadbyte && $leadbyte <= 0xFC))) {
			return -1;
		}
		$trailbyte = ($cp932 & 0xff);
		if (!((0x40 <= $trailbyte && $trailbyte <= 0x7E) ||
			  (0x80 <= $trailbyte && $trailbyte <= 0xFC))) {
			return -1;
		}
		$inbuf = sprintf("%c%c", $leadbyte, $trailbyte);
	}
	mb_substitute_character(0xfeff);
	$unicode = 0;
	$outbuf = mb_convert_encoding($inbuf, "UTF-16BE", "CP932");
	if (!(strlen($outbuf) == 2 && $outbuf != "\xfe\xff")) {
		return -1;
	}
	return hexdec(bin2hex($outbuf));
}

function main() {
	for ($cp932 = 0; $cp932 <= 0xffff; ++ $cp932) {
		$unicode = decode($cp932);
		if (0 <= $unicode) {
			if ($cp932 < 0x100) {
				printf("0x%02X\t", $cp932);
			} else {
				printf("0x%04X\t", $cp932);
			}
			printf("0x%04X\n", $unicode);
		}
	}
}

main();
